Introduce performance tests and multi pipeline PoCs for metrics (#778)
kyma-project/community: concepts/observability-strategy/configurable-tracing/pocs/pipeline-investigations/README.md

# Trace Performance Test

The test covers 3 aspect of default base installation and configuration.

- Test average throughput of current installation end-to-end.
- Test queuing and retry capabilities of installation with simulated backend outages.

> - Test survivability of the current installation under high load. The collector should not run out of memory in any case. Instead, it should refuse incoming data until memory available again.

Reviewer: ```suggestion
- Test survivability of the current installation under high load. The collector should not run out of memory in any case. Instead, it should refuse incoming data until memory is available again.
```

> **NOTE:**
> The Opentelemetry Collector test environment is configured with OTLP format over the gRPC protocol. All KPIs found in this test will vary with other formats and protocols.

## Test Application

> The [Opentelemetry load generator](https://github.tools.sap/huskies/otel-load-generator) is developed by Huskies. The application written is in go and use the Opentelemetry SDK version v1.11.1.

Reviewer: ```suggestion
The [Opentelemetry load generator](https://github.tools.sap/huskies/otel-load-generator) is developed by Kyma's Telemetry team. The application written is in Go and uses the Opentelemetry SDK version v1.11.1.
```

The load generator produces for each generated trace `11 spans` (one root and 10 child spans) and each spans has `40` attributes each `64 Byte` (except root span) long random data, this will simulate maximum size of expected span data,
generated data send in batches over `gRPC` and each batch will contain `512` spans, the test data can be generated concurrently to increase load on targeted collector, this can can be configured with start parameter `-c`.

## Test Environment

- A Kyma cluster deployed with Kyma version 2.9.x
- Opentelemetry Collector version 0.60.0 deployed by the Kyma Telemetry Controller with following configuration:
  - 1GiB Memory
  - 1 CPU
  - Memory hard limit configured 75% of total memory, memory spike limit configured to the 10% of total available memory.
  - Retry backoff limit 30 seconds and maximum retry time to 300 seconds.

>   - Bath size 512 spans per batch, maximum batch queue size 512 batches.

Reviewer: what bath? Is it batch?

```suggestion
  - Batch size 512 spans per batch and maximum batch queue size 512 batches
```

| :--: |
| Fig. 3 Refused spans |

- `Fig. 3` show the behaviour of memory limiter in case of high load and high memory utilization. Memory limiter kick in when the memory utilization get above defined `hard limit` and start refuse incoming data to avoid OOM situation.

### Retry and Queuing Test

This scenario tests an outage of the backend. In case of a backend outage data should be queued for max `300 seconds` and retried to export before data is dropped from queue.

| ![Batch Queue](assets/queue_test_long.png) |
| :--: |
| Fig. 4 Batch queue |

The `Fig. 4` shows the queue size is going up in time and going down approximately every 5 minutes. This is because queued data will be dropped when could not be exported successfully after 300 second (5 Minutes).

| ![Failed Exports](assets/failed_send_spans_long.png) |
| :--: |
| Fig. 5 Failed exports |

The `Fig. 5` show unsuccessful export attempts during test execution.

### Test with CLS Backend

The throughput test runs over one hour to measure average throughput of span ingestion. In the peak times, no span should be dropped or refused by the pipeline and overall resource consumption should stay moderate.

> The test uses a CLS instance without any configuration changes as backend. The test load generator is started with 2 concurrent producers, which produce on average `1550 spans/sec`.

Reviewer: ```suggestion
As backend, the test uses a CLS instance without any configuration changes. The test load generator is started with 2 concurrent producers, which produce on average 1550 spans/sec.
```

| ![Accepted Spans](assets/accepted_spans_cls.png) |
| :--: |
| Fig. 6 Accepted spans |

| ![Send Exports](assets/send_spans_cls.png) |
| :--: |
| Fig. 7 Send exports |

| ![Queue Size](assets/queue_size_cls.png) |
| :--: |
| Fig. 8 Queue size |

| ![CPU Utilization](assets/cpu_cls.png) |
| :--: |
| Fig. 9 CPU Utilization |

| ![Memory Utilization](assets/memory_cls.png) |
| :--: |
| Fig. 10 Memory Utilization |

> The CLS instance became approximately after 4 minutes unavailable, which can be observed from `Fig. 8`. After 4 minutes of load test, the queue size starts to grow because spans can't be successfully exported.

Reviewer: ```suggestion
The CLS instance became unavailable after approximately 4 minutes, which can be observed from `Fig. 8`. After 4 minutes of load test, the queue size starts to grow because spans can't be successfully exported.
```

go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).onTemporaryFailure
 go.opentelemetry.io/collector@v0.60.0/exporter/exporterhelper/queued_retry.go:176
go.opentelemetry.io/collector/exporter/exporterhelper.(*retrySender).send
 go.opentelemetry.io/collector@v0.60.0/exporter/exporterhelper/queued_retry.go:411
go.opentelemetry.io/collector/exporter/exporterhelper.(*tracesExporterWithObservability).send
 go.opentelemetry.io/collector@v0.60.0/exporter/exporterhelper/traces.go:134
go.opentelemetry.io/collector/exporter/exporterhelper.(*queuedRetrySender).start.func1
 go.opentelemetry.io/collector@v0.60.0/exporter/exporterhelper/queued_retry.go:206
go.opentelemetry.io/collector/exporter/exporterhelper/internal.(*boundedMemoryQueue).StartConsumers.func1
 go.opentelemetry.io/collector@v0.60.0/exporter/exporterhelper/internal/bounded_memory_queue.go:61
```

### Throughput Test

> The throughput test ran over one hour to measure average throughput of span ingestion, in the peak times no span should be dropped or refused by the pipeline and overall resource consumption should stay moderate.

Reviewer: ```suggestion
The throughput test ran over one hour to measure average throughput of span ingestion. In the peak times, no span should be dropped or refused by the pipeline, and overall resource consumption should stay moderate.
```

> As a backend, Kyma's integrated Jaeger instance is used. To avoid Jaeger getting a bottleneck in the end-to-end pipeline, the Jaeger instance is configured to use `1.5GiB` memory. The in-memory trace storage of Jaeger instance is reduced to the maximum of `500` traces.

Reviewer: ```suggestion
As a backend, Kyma's integrated Jaeger instance is used. To avoid Jaeger becoming a bottleneck in the end-to-end pipeline, the Jaeger instance is configured to use 1.5GiB memory. The in-memory trace storage of Jaeger instance is reduced to the maximum of 500 traces.
```

To disclose the impact of the backend on the KPIs, a second OpenTelemetry Collector is deployed and used as OTLP gRPC backend. The same test is repeated on this setup.
The KPIs are the same as the setup with Jaeger backend.

| ![Send Exports](assets/send_spans_throughput.png) |
| :--: |
| Fig. 11 Send exports |

| ![Accepted Spans](assets/accepted_spans_throughput.png) |
| :--: |
| Fig. 12 Accepted spans |

| ![CPU Utilization](assets/cpu_throughput.png) |
| :--: |
| Fig. 13 CPU Utilization |

| ![Memory Utilization](assets/memory_throughput.png) |
| :--: |
| Fig. 14 Memory Utilization |

> The `Fig. 12` shows the amount of spans successfully ingested and the `Fig. 11` shown the amount of spans successfully exported to the configured backend. Both charts `Fig. 12` and `Fig. 11` show the same value of `~16000000` spans, which indicate that all ingested data successfully exported.

Reviewer: ```suggestion
`Fig. 12` shows the amount of spans successfully ingested. 
`Fig. 11` shows the amount of spans successfully exported to the configured backend.
Both charts, `Fig. 12` and `Fig. 11`, show the same value of ~16000000 spans, which indicate that all ingested data has been successfully exported.
```

> This measurement result in an average of `4300` span per seconds.

Reviewer: ```suggestion
This measurement results in an average of 4300 spans per seconds.
```

> The `Fig. 14` show the memory utilization of the OpenTelemetry Collector instance. The measured value stay around 90% of total available memory.

Reviewer: ```suggestion
`Fig. 14` shows the memory utilization of the OpenTelemetry Collector instance. The measured value stays around 90% of total available memory.
```

> `Fig. 13` shows the CPU utilization of OpenTelemetry Collector instance. The measurement shows that the available CPU shares are fully utilized.

Reviewer: why first Fig 14 and then Fig 13?

The OpenTelemetry Collector version used in this test has some limitations with the gRPC client. The used gRPC client has a default message size of 4194304 bytes, which prevents exporting bigger batch sizes to the backend. 

> **NOTE:** Repeat this test with bigger batch sizes when the size limit is fixed. See the open issue [Max Export Size for Metrics/Logs/Traces #2772](https://github.com/open-telemetry/opentelemetry-specification/issues/2772).

## Multi Pipeline Test

The purpose of this test is to understand the behavior of the trace pipeline gateway with multiple pipelines.

### Test Environment

> - Telemetry tracing gateway used the following config

Reviewer: ```suggestion
- Telemetry tracing gateway with the following config:
```

                receivers:
                    - opencensus
                    - otlp
                processors:
                    - memory_limiter
                    - k8sattributes
                    - filter
                    - resource
                    - batch
                exporters:
                    - otlp
                    - logging
            traces/2:
                receivers:
                    - opencensus
                    - otlp
                processors:
                    - memory_limiter
                    - k8sattributes
                    - filter
                    - resource
                    - batch
                exporters:
                    - otlp/2
                    - logging
        telemetry:
            metrics:
                address: ${MY_POD_IP}:8888
            logs:
                level: info
        extensions:
            - health_check
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: telemetry-trace-collector
  name: telemetry-trace-collector
  namespace: kyma-system
```
- A Kyma cluster deployed with Kyma version 2.11.x
- Custom Kyma OpenTelemetry Collector image based on version 0.74.0 

> - Backend [mock-trace-backend](https://github.tools.sap/huskies/knowledge-hub/tree/master/docs/observability/mock-trace-backend-poc), and two instances deployed for each configured pipeline on different Namespaces. Deployment used for both backend deployment

Reviewer: What does this sentence mean?  "Deployment used for both backend deployment"

    metadata:
      labels:
        app: mock-trace-receiver
    spec:
      volumes:
        - name: collector-config
          configMap:
            name: mock-trace-receiver
        - name: data
          emptyDir: {}
      securityContext:
        fsGroup: 101
      containers:
      - image: otel/opentelemetry-collector-contrib:0.70.0
        name: otel-collector
        securityContext:
          runAsUser: 101
        volumeMounts:
        - name: collector-config
          mountPath: /etc/collector
        - name: data
          mountPath: /traces
        args:
        - --config=/etc/collector/config.yaml
      - image: nginx
        name: web
        volumeMounts:
        - name: data
          mountPath: /usr/share/nginx/html
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: mock-trace-receiver
  name: mock-trace-receiver
  namespace: trace-receiver
spec:
  ports:
  - name: http-otlp
    port: 4317
    protocol: TCP
    targetPort: 4317
  - name: grpc-otlp
    port: 4318
    protocol: TCP
    targetPort: 4318
  - name: export-http
    port: 8080
    protocol: TCP
    targetPort: 80
  selector:
    app: mock-trace-receiver
  type: LoadBalancer
```

> - Load generator used pipeline for tests

Reviewer: what does this sentence mean?

```
- A Kyma cluster deployed with Kyma version 2.11.x
- Custom Kyma OpenTelemetry Collector image based on version 0.74.0 

> - `mock-trace-backend` had two instances deployed for each configured pipeline in different namespaces.

Reviewer: ```suggestion
- `mock-trace-backend` had two instances deployed for each configured pipeline in different Namespaces.
```
